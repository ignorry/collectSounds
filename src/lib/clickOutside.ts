/**
 * Listen to clicks outside of the element
 * @param {HTMLElement} elem
 * @param {Function} callback 
 * @returns {void}
 */
export const clickOutside = ( elem: HTMLElement, callback: Function ): void =>
  document.addEventListener( 'click', (e) => {
    const target: Node = e.target as Node;
    
    if ( target !== elem && ! elem.contains( target ) )
      callback();
  });