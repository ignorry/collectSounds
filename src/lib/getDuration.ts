/**
 * get duration from ms in readable format (h:mm:ss)
 * @param {number} duration 
 * @returns {string}
 */
export const getDurationFromMs = ( duration: number ): string =>
  `${ duration/3600000 > 1 ? `${ Math.floor( duration/3600000 ) }:` : ''
  }${ duration/60000 > 1 ? `${ Math.floor( duration%3600000/60000 ) }:` : ''
  }${ `${ Math.floor( duration%60000/1000 ) }` }`