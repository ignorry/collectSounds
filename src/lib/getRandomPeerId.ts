/**
 * generates 64 characters id starting with 'collectS'
 * @returns { string }
 */
export const getRandomPeerId = () => {
  let result = '';
  const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  for ( let i = 0; i < 56; i++ ) {
    result += characters.charAt( Math.floor( Math.random() * characters.length ) );
  }

  return 'collectS' + result;
}
