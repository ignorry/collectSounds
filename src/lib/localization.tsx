import React, { ReactNode, useState, useEffect } from "react";
import { IntlProvider } from "react-intl";

export type Lang = 'en' | 'ru';
export type Page = 
  'search' |
  'saved' |
  'queue' |
  'settings' |
  'video' |
  'playlist' |
  'channel' |
  'synchronization';

/**
 * import specific translation file
 * @param {Page} page 
 * @param {Lang} lang 
 */
export const loadMessages = ( page: Page, lang: Lang ) => import( `../translations/${page}/${lang}.json` );

/**
 * get user's language
 * @returns {Lang} lang code (eg 'en', 'ru')
 */
export const getLang = (): Lang => {
  const code = navigator.language.slice( 0, 2 );

  if ( code === 'ru' ) return code;

  return 'en';
}

/**
 * @typedef Props
 * @prop {string} file - translation file
 * @prop {ReactNode} children
 */
export type Props = {
  file: Page;
  children?: ReactNode;
  skeleton?: ReactNode;
}

/**
 * wrap component with TranslationProvider
 * @param {Props} props
 */
export const LangProvider: React.FC<Props> = ( props: Props ) => {
  const locale = getLang();
  const [messages, setMessages] = useState( null );

  useEffect(() => {
    loadMessages( props.file, locale ).then( setMessages );
  }, [locale]);

  return messages ? (
    <IntlProvider locale={locale} messages={messages}>
      { props.children }
    </IntlProvider>
  ) : <> { props.skeleton } </>;
}