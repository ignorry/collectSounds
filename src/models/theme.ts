export enum ThemeEnum  {
  light = 'light',
  dark = 'dark'
}

// all sizes in rem
export type Theme = {
  iconSize: number,
  rangeSize: number,
  spinnerSize: number,
  contentWidth: number,
  colors: {
    main: string,
    bg?: string,
    bgSecondary?: string,
    font?: string,
    darkBg?: string,
  },
  decorative: { // in px
    borderRadius: number;
    padding: number;
  },
  gaps: {
    smallest: number,
    small: number,
    big: number,
  },
  thumbnail: {
    width: number,
    height: number,
    infoColor: string,
    infoWidth: 4,
  },
  error: {
    backgroundColor: string,
    width: number,
  },
  media: {
    small: string,
  },
  font: {
    primary: {
      size: number,
      opacity: number,
    },
    secondary: {
      size: number,
      opacity: number,
    },
  },
  durations: {
    ms300: number,
  },
  order: {
    footer: number,
    modal: number,
  },
}