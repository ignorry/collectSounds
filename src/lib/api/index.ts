import { 
  fetchVideo,
  fetchPlaylist,
  fetchPlaylistItems,
  fetchChannel,
  fetchSearch
} from "./fetch";
import {
  parseVideo,
  parsePlaylist,
  parseChannel
} from "./parseApiResponse";

import { Video, Playlist, Channel, Content } from "../../models/content";

/**
 * fetch video by id
 * @param {string} id - video id
 * @returns {Promise<Array<Video>>} 
 */
export const getVideoById = async ( id: string ): Promise<Array<Video>> => {
  const res = await fetchVideo({
    part: 'snippet,contentDetails',
    id
  });

  return parseVideo( res );
}

/**
 * fetch playlist by id
 * @param {string} id - playlist id
 * @returns {Promise<Array<Playlist>>}
 */
export const getPlaylistById = async ( id: string ): Promise<Array<Playlist>> => {
  const res = await fetchPlaylist({
    part: 'snippet',
    id
  });

  return parsePlaylist( res );
}

/**
 * fetch channel by id
 * @param {string} id - channel id
 * @returns {Promise<Array<Channel>>}
 */
export const getChannelById = async ( id: string ): Promise<Array<Channel>> => {
  const res = await fetchChannel({
    part: 'snippet',
    id
  });

  return parseChannel( res );
}

/**
 * fetch playlist's videos
 * @param {string} id - playlist id
 * @returns {Promise<Array<Video>>}
 */
export const getPlaylistVideos = async ( id: string ): Promise<Array<Video>> => {
  const res = await fetchPlaylistItems({
    part: 'id',
    playlistId: id,
  });

  const ids: Array<string> = res.items.map( item => item.id );

  return getVideoById( ids.join( ',' ) );
}

/**
 * fetch channel's videos
 * @param {string} id - channel id
 * @returns {Promise<Array<Video>>}
 */
export const getChannelVideos = async ( id: string ): Promise<Array<Video>> => {
  const res = await fetchSearch({
    part: 'id',
    channelId: id,
    order: 'date',
  })

  const ids: Array<string> = res.items.filter( item => item.id.kind === 'video' ).map( item => item.id.videoId );

  return getVideoById( ids.join( ',' ) );
}

/**
 * fetch channel's playlists
 * @param {string} id - channel id
 * @returns {Promise<Array<Playlist>>}
 */
export const getChannelPlaylists = async ( id: string ): Promise<Array<Playlist>> => {
  const res = await fetchPlaylist({
    part: 'snippet',
    channelId: id,
  });

  return parsePlaylist( res )
}

/**
 * @typedef SearchFilters
 * @prop {string} order - specifies the method that will be used to order resources in the API response
 * @prop {string} type - restricts a search query to only retrieve a particular type of resource
 * @prop {string} videoDuration - filters video search results based on their duration
 * @prop {string} saveSearch - indicates whether the search results should include restricted content as well as standard content
 */
export type SearchFilters = {
  order?: 'date' | 'rating' | 'relevance' | 'title' | 'videoCount' | 'viewCount';
  type?: 'video' | 'playlist' | 'channel',
  videoDuration?: 'any' | 'long' | 'medium' | 'short';
  saveSearch?: 'moderate' | 'none' | 'strict'
}

/**
 * fetch search results with specified filters
 * @param {SearchFilters} filters
 * @returns {Promise<Array<Content>>}
 */
export const getSearch = async ( filters: SearchFilters ): Promise<Array<Content>> => {
  const res = await fetchSearch({...filters,
    part: 'id',
    maxResults: 10,
  });

  const videoIds: Array<string> = [];
  const playlistIds: Array<string> = [];
  const channelIds: Array<string> = [];
  const ids: Array<string> = res.items.map( item => {
    if ( item.id.kind === 'video' ) {
      videoIds.push( item.id.videoId );
      return item.id.videoId;
    } else if ( item.id.kind === 'playlist' ) {
      playlistIds.push( item.id.playlistId );
      return item.id.playlistId;
    }

    channelIds.push( item.id.channelId );
    return item.id.channelId;
  });

  const videos: Array<Video> = await getVideoById( videoIds.join( ',' ) );
  const playlists: Array<Playlist> = await getPlaylistById( playlistIds.join( ',' ) );
  const channels: Array<Channel> = await  getChannelById( channelIds.join( ',' ) );

  // ids array needed to save order of contents

  return ids.map( item => {
    if ( videos[0] && videos[0].id === item ) return videos.shift();
    if ( playlists[0] && playlists[0].id === item ) return playlists.shift();
    return channels.shift();
  });
}