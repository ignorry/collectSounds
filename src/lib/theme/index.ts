import { DefaultTheme } from "styled-components";
import { ThemeEnum } from "../../models/theme";
import { baseTheme } from "./baseTheme";

export const lightTheme: DefaultTheme = {
  ...baseTheme,
  type: ThemeEnum.light,

  colors: {
    ...baseTheme.colors,
    main: '#00bb88',
    bg: '#ffffff',
    bgSecondary: '#d8d8d8',
    font: '#000000',
    darkBg: '#d9d9d9'
  },
}

export const darkTheme: DefaultTheme = {
  ...baseTheme,
  type: ThemeEnum.dark,

  colors: {
    ...baseTheme.colors,
    bg: '#000000',
    bgSecondary: '#2a2a2a',
    font: '#ffffff',
    darkBg: '#121212'
  },
}

/**
 * get theme depending on isDark parameter
 * @param {boolean} isDark 
 * @returns {DefaultTheme}
 */
export const getTheme = ( isDark: boolean ): DefaultTheme => isDark ? darkTheme : lightTheme;