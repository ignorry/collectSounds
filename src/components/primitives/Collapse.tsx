import React, { ReactNode, useEffect, useRef, useState } from "react";
import styled from "styled-components";
import { clickOutside } from "../../lib/clickOutside";

import ToggleButton from "./ToggleButton";

export const Div = styled.div<{ display: boolean }>`
  display: flex;
  flex-direction: column;
  gap: ${ props => props.display ? ({ theme }) => `${ theme.gaps.small }rem` : 0 };
  padding: ${ ({ theme }) => `${ theme.decorative.padding }px`};
  text-align: center;
`;

export const Content = styled.div<{ display: boolean }>`
  height: ${ props => props.display ? 'auto' : 0 };
  overflow: hidden;
`;

/**
 * @typedef Props
 * @prop {string} label - display in left of active value
 * @prop {any} [children] - react prop
 * @prop {string} [icon] - name of icon to be displayed in button
 */
export type Props = {
  label: string,
  children?: any,
  icon?: string,
}

const Collapse: React.FC<Props> = ( props: Props ) => {
  const [ display, setDisplay ] = useState<boolean>( false );
  const containerRef: React.MutableRefObject<undefined | HTMLDivElement> = useRef();

  const toggleDisplay = () => setDisplay( ! display );

  useEffect( () => {
    clickOutside( containerRef.current , () => setDisplay( false ) )
  }, []);

  return (
    <Div display={ display } ref={ containerRef }>
      <ToggleButton pic={ props.icon } text={ props.label } callback={ toggleDisplay } active={ display } />

      <Content display={ display }>
        { props.children }
      </Content>
    </Div>
  );
}

export default Collapse;