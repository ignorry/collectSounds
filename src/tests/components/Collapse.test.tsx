import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ThemeProvider } from "styled-components";
import { getTheme } from "../../lib/theme";
import GlobalStyles from "../../lib/styled/global";

import Collapse from "../../components/primitives/Collapse";

describe( 'Collapse component', () => {
  it( 'display button and content', () => {
    const result = render(
      <ThemeProvider theme={ getTheme( true ) } >
        <Collapse label="label" children={<span>child</span>} icon="logo"/>
        <GlobalStyles/>
      </ThemeProvider>
    );

    const button = result.container.querySelector( 'button' );
    const content = result.container.querySelector( 'span' );

    expect( button ).toBeDefined();
    expect( content.textContent ).toBe( 'child' );
  });

  it( 'displays content when expanded', async () => {
    const user = userEvent.setup()

    const result = render(
      <ThemeProvider theme={ getTheme( true ) } >
        <Collapse label="label" children={<span>child</span>} icon="logo"/>
        <GlobalStyles/>
      </ThemeProvider>
    );

    await user.click( screen.getByRole('button') )

    const content = result.container.querySelector( 'span' ).parentElement;

    expect( window.getComputedStyle( content ).height ).toBe( 'auto' );
  });

  it( 'matches to snapshot', () => {
    const result = render(
      <ThemeProvider theme={ getTheme( true ) } >
        <Collapse label="label" children={<p>child</p>} icon="logo"/>
        <GlobalStyles/>
      </ThemeProvider>
    );

    expect( result ).toMatchSnapshot();
  });
});