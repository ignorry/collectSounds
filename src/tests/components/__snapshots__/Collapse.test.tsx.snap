// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapse component matches to snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-hKMtZM DcwtX"
      >
        <button
          class="sc-dkzDqf fcjXBB"
        >
          <svg
            class="sc-bczRLJ bygNPW"
          >
            <use
              href="icons.svg#logo"
            />
          </svg>
          <p
            class="sc-gsnTZi cBa-dYy"
          >
            label
          </p>
        </button>
        <div
          class="sc-eCYdqJ gJLhFV"
        >
          <p>
            child
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-hKMtZM DcwtX"
    >
      <button
        class="sc-dkzDqf fcjXBB"
      >
        <svg
          class="sc-bczRLJ bygNPW"
        >
          <use
            href="icons.svg#logo"
          />
        </svg>
        <p
          class="sc-gsnTZi cBa-dYy"
        >
          label
        </p>
      </button>
      <div
        class="sc-eCYdqJ gJLhFV"
      >
        <p>
          child
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
